[#14] Feat: TypeScript 적용

sageherb opened the pull request:
## 개요

이슈 번호: #14 

타입스크립트 설정 및 적용

## 작업 내용

- [x] 타입스크립트, `ts-loader` 설치 및 `tsconfig.json` 설정
- [x] 타입스크립트 웹팩 설정
- [x] `types` 폴더 생성 및 `component.d.ts`, `module.d.ts` 파일 작성
- [x] 기존 js 파일 ts 파일로 변경 및 타입스크립트 적용

## 기타

sageherb committed:
Chore: TypeScript 설정

- typescript, ts-loader 설치
- tsconfig.json 설정
- webpack typescript 설정
- index.js -> index.ts로 변경

sageherb committed:
Chore: tsconfig.json 및 gitignore 변경

sageherb committed:
Feat: Custom Element declare

sageherb committed:
Refactor: App TypeScript 적용

sageherb committed:
Refactor: formatDate TypeScript 적용

sageherb committed:
Chore: tsconfig 수정 및 types 폴더 생성 후 d.ts 파일 생성

- tsconfig.json 파일에 typeRoots 옵션 추가
- types 폴더에 module.d.ts 파일 생성
- global.d.ts -> component.d.ts 수정 및 이동

sageherb committed:
Feat: Footer 타입스크립트로 변경

sageherb committed:
Feat: FooterIconList TypeScript 적용

- svg 모듈 declare 추가
- 기존 코드 리터럴 템플릿으로 변경

sageherb committed:
Feat: FooterList TypeScript 적용

- 기존 코드 리터럴 템플릿으로 변경
- 타입스크립트 적용

sageherb committed:
Feat: Header TypeScript 적용 및 커스텀 이벤트 변경

sageherb committed:
Refactor: axios TypeScript 적용

sageherb committed:
Feat: axios articleList TypeScript 적용

- articleList 타입스크립트 적용
- types.d.ts에 Article 타입 생성

sageherb committed:
Feat: ArticleList TypeScript 적용

sageherb committed:
Refactor: Header 이벤트 콜백 분리

f-lab-liv commented on src/types/component.d.ts:
```diff
@@ -0,0 +1,19 @@
+import FooterList from "@/components/FooterList/FooterList";
+import FooterIconList from "@/components/FooterIconList/FooterIconList";
+
+declare global {
+  interface HTMLElementTagNameMap {
+    "blog-header": Header;
+    "blog-section": Section;
+    "article-list": ArticleList;
+    "article-item": ArticleItem;
+    "blog-footer": Footer;
+    "footer-list": FooterList;
+    "footer-icon-list": FooterIconList;
+  }
+}
```
👍

f-lab-liv commented on src/types/module.d.ts:
```diff
@@ -0,0 +1,9 @@
+declare module "*.html" {
+  const content: string;
+  export default content;
+}
```
~~이건 어떤 부분 때문에 필요하게 된 걸까요?~~
html import해서 template 사용하는 부분이 있었죠.. 확인했습니다.

sageherb replied:
각 컴포넌트에서 template.html 파일을 모듈로 인식하기 위해서 설정했습니다!

f-lab-liv commented on src/components/ArticleList/ArticleList.ts:
```diff
@@ -2,60 +2,71 @@ import template from "./template.html";
 import { getTechArticleList, getDesignArticleList } from "@/api/articleList";
+import { Article } from "@/types/types";
 
-const TITLE_MAP = {
+const TITLE_MAP: { [key: string]: string } = {
```
Record를 활용할 수 있을까요?

sageherb replied:
Record로 변경하겠습니다!
멘토 님 이 부분 질문이 있습니다!
해당 컴포넌트에서 사용하는 상수를 클래스 내 static 프로퍼티로 두는 게 좋을지, 지금처럼 컴포넌트 외부의 상수로 그냥 둘지 고민스러웠는데 어떤 게 더 자연스러울까요?

f-lab-liv replied:
class 외부의 상수로 두어도 무방할 것 같습니다.

f-lab-liv commented on src/api/articleList.ts:
```diff
@@ -0,0 +1,41 @@
+    return data;
+  } catch (error) {
+    if (axios.isAxiosError<{ message: string }>(error)) {
+      console.log(error.response?.data.message);
```
에러 로깅이 목적이라면 console.log보다는 console.error가 적절해보입니다.

sageherb replied:
수정하겠습니다!

sageherb committed:
Update: console.log console.error로 변경

sageherb committed:
Update: 상수 타입 Record로 변경

sageherb committed:
Feat: removeEventListener 추가

sageherb committed:
Feat: ArticleItem TypeScript 적용

- 타입스크립트로 코드 수정
- html 템플릿에서 리터럴 템플릿으로 수정
- ArticleItem 이벤트 삭제 및 ArticleList에서 이벤트 위임

sageherb committed:
Feat: ArticleItem 클릭 이벤트 구현

sageherb committed:
Feat: Section TypeScript 적용

sageherb committed:
Fix: Router 확장자 변경

f-lab-liv commented on src/components/Section/Section.ts:
```diff
@@ -0,0 +1,26 @@
+    this.configureRouting();
+  }
+
+  configureRouting(): void {
+    this.router.addRoute("/", () => ArticleListPage("tech"));
+    this.router.addRoute("/tech", () => ArticleListPage("tech"));
+    this.router.addRoute("/design", () => ArticleListPage("design"));
+    this.router.addRoute("/article/:id", (params) => ArticlePage(params));
+    this.router.setNotFoundPage(() => NotFoundPage());
+  }
```
👍

sageherb committed:
Feat: Router TypeScript 적용 및 pageNavigation 이벤트 수정

sageherb committed:
Feat: 파일 이름 변경 및 pageNavigation declare 추가

sageherb committed:
Feat: parseTemplate 생성 및 ArticleList, Footer, Header 컴포넌트 Template 생성 시 호출

sageherb committed:
Feat: a11y-hidden 추가

sageherb committed:
Refactor: axios 에러 처리 리팩토링

sageherb committed:
Refactor: Page TypeScript 적용

f-lab-liv commented on src/api/articleList.ts:
```diff
@@ -1,41 +1,25 @@
-      method: "GET",
-      url: "data/mockTechArticleList.json",
-    });
+  const response: AxiosResponse = await request({
```
request의 return type에 의해서 자연스럽게 추론되게 하는 것이 더 좋아보입니다.
추가적으로, Generic을 활용하여 타입을 추론하는 것은 어떨까요?

f-lab-liv commented on src/pages/ArticleListPage.ts:
```diff
@@ -1,7 +1,7 @@
 
-const ArticleListPage = (path = "tech") => {
-  const articleList = new ArticleList(path);
+const ArticleListPage = (path: string = "tech"): ArticleList => {
```
path의 type정의를 string 대신 다르게 할 수 있을까요?

f-lab-liv commented on src/pages/NotFoundPage.ts:
```diff
@@ -1,5 +1,5 @@
-const NotFoundPage = () => {
-  const div = document.createElement("div");
+const NotFoundPage = (): HTMLDivElement => {
+  const div: HTMLDivElement = document.createElement("div");
```
타입을 선언하지 않고, 추론되게 하는 것이 좋을 것 같습니다. ~~반면 함수의 return type은 유지하는 것이 좋은데, 이유도 생각해보시면 좋을 것 같습니다.~~
제가 잠깐 잘못 생각했네요. return type 또한 반드시 필요하지 않다면 typescript의 추론에 맞기는 것이 좋을 것 같네요.

f-lab-liv commented on src/utils/parseTemplate.ts:
```diff
@@ -0,0 +1,11 @@
+export const parseTemplate = (template: string): DocumentFragment => {
+  const parsedTemplate = new DOMParser()
+    .parseFromString(template, "text/html")
+    .querySelector("template");
+
+  if (!parsedTemplate) {
+    throw new Error("Template Error");
+  }
+
+  return parsedTemplate.content;
+};
```
👍

f-lab-liv commented on src/api/axios.ts:
```diff
@@ -0,0 +1,26 @@
+      data,
+    });
+    return response;
+  } catch (error) {
+    throw new Error("Request Error");
+  }
```
큰 의미가 없는 에러핸들링같습니다. 오히려, axios에서 발생한 에러를 모두 `new Error("Request Error");` 로 바꿔버리기 때문에 실제로 어떤 에러가 발생했는지 어려워지는 단점이 있어보입니다.

f-lab-liv commented on src/components/ArticleList/ArticleList.ts:
```diff
@@ -0,0 +1,83 @@
+import { getTechArticleList, getDesignArticleList } from "@/api/articleList";
+import { Article } from "@/types/types";
+
+const TITLE_MAP: Record<string, string> = {
```
https://github.com/f-lab-edu/vanilla-spa-clone/pull/15/files#r1537696151
와 동일한 피드백입니다.
Record의 Generic으로 들어가는 첫번째 type을 수정할 수 있을까요?

f-lab-liv commented on src/components/ArticleList/ArticleList.ts:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  render(): void {
+    if (!this.template) return;
```
불필요한 early return 같습니다!

f-lab-liv commented on src/router/router.ts:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  stop(): void {
+    window.removeEventListener("popstate", () => this.checkRoutes());
+    window.removeEventListener("pageNavigation", (event) =>
+      this.navigate(event.detail.path)
+    );
```
여기서 removeEventListener에 들어가는 callback들은 addEventListner에서 추가해주는 callback과 동일한 것이 맞을까요?

sageherb replied:
콜백 함수가 각자의 함수로 지정되어 동일하지 않은데 생각을 못했습니다! 수정하겠습니다!

f-lab-liv commented on src/components/ArticleList/ArticleList.ts:
```diff
@@ -0,0 +1,83 @@
+    this.getArticles();
+  }
+
+  disconnectedCallback(): void {
+    this.removeEventListener("articleListLoaded", () => this.render());
+    this.removeEventListener("click", this.handleNavigationClick);
+  }
```
이벤트 리스너 해제가 꼭 필요할까요? router와 다른 점은 무엇일까요?

sageherb replied:
router의 경우 window에 이벤트 리스너가 설정됐는데, 이 경우 컴포넌트에 설정되었고 컴포넌트가 disconnected 될 때 자동으로 사라져서 불필요할 것 같습니다! 수정하겠습니다!

f-lab-liv commented on src/components/Header/Header.ts:
```diff
@@ -0,0 +1,43 @@
+    });
+  }
+
+  disconnectedCallback(): void {
+    const aList: NodeListOf<HTMLAnchorElement> =
+      this.querySelectorAll("a[data-navigation]");
+
+    aList.forEach((a) => {
+      a.removeEventListener("click", this.handleNavigationClick);
+    });
+  }
```
https://github.com/f-lab-edu/vanilla-spa-clone/pull/15/files#r1537727623
와 동일합니다.

f-lab-liv commented on webpack.common.js:
```diff
@@ -3,9 +3,9 @@ const HtmlWebpackPlugin = require("html-webpack-plugin");
+  entry: path.resolve(__dirname, "src/index.ts"),
   resolve: {
-    extensions: [".js"],
+    extensions: [".ts", ".js"],
```
ts로의 전환이 모두 완료되면 `.js`는 빠져도 될 것 같습니다!

sageherb committed:
Fix: 불필요한 Early Return 제거

sageherb committed:
Fix: Page 타입 수정

sageherb committed:
Fix: 불필요한 js 타입 제거

sageherb committed:
Fix: request 함수 제네릭 추가

sageherb committed:
Fix: ArticlePath 타입 정의 및 사용

sageherb committed:
Fix: Page 컴포넌트 타입 추론 사용

sageherb committed:
Fix: ArticlePath 타입 사용

sageherb committed:
Fix: 이벤트 핸들러 추가

sageherb committed:
Fix: 불필요한 코드 삭제 및 ArticlePath 타입 설정

sageherb committed:
Fix: webpack 설정 변경

sageherb committed:
Fix: Page type 선언 및 this 바운딩

sageherb committed:
Feat: 에러 페이지 생성 및 Article 요청 에러 처리 수정

sageherb committed:
Fix: ArticlePage에서 params 없을 때 및 Not Found 페이지 없을 때 처리

f-lab-liv commented on src/router/router.ts:
```diff
@@ -0,0 +1,103 @@
+  private routes: Routes[] = [];
+  private currentPath: string | null = null;
+  private notFoundPage: Page | null = null;
+  private boundHandlePopState = this.handlePopState.bind(this);
+  private boundHandlePageNavigation = this.handlePageNavigation.bind(this);
```
👍 👍

f-lab-liv replied:
혹시 참고하셨던 링크가 있었을까요?

sageherb replied:
https://stackoverflow.com/questions/61148618/bind-this-in-class-method

class this binding 검색해서 스택 오버 플로우 글 참고해서 적용했습니다!

f-lab-liv replied:
제가 의도한 방향으로 잘 수정해주신 것 같네요. this binding에 대해서도 한 번 더 이해할 수 있는 기회가 된 것 같네요!

참고로 react-query 같은 경우에는 https://github.com/TanStack/query/blob/0a2b1ba18569aa71ef2e2de6388180aa34470df7/packages/query-core/src/queryObserver.ts#L87
이런 방법을 사용했더라고요.

f-lab-liv commented on src/components/ArticleList/ArticleList.ts:
```diff
@@ -27,8 +27,14 @@ export default class ArticleList extends HTMLElement {
+      window.dispatchEvent(
+        new CustomEvent("pageNavigation", { detail: { path: "/error" } })
+      );
+    }
```
👍

f-lab-liv commented on src/pages/ArticlePage.ts:
```diff
@@ -0,0 +1,14 @@
+import NotFoundPage from "./NotFoundPage";
+
+const ArticlePage = (params?: Record<string, string>) => {
+  if (!params) {
+    return NotFoundPage();
+  }
```
사실 ArticlePage는 param이 꼭 필요한 페이지라고 볼 수 있는데 현재 Page type의 정의로는 어쩔 수 없이 optional을 처리해줘야 하는 부분이 조금 있네요 😢 
이런 부분은 지금 수정할 필요는 없어보이나, 나중에 type safe한 라우팅을 할 수 있는 방법에 대해 조금 더 고민해보아도 좋을만한 포인트 같습니다.

f-lab-liv reviewed (approved):
LGTM 입니다!
